Segment-Optimizely-X/Classes/SEGOptimizelyXIntegration.m

```diff
@@ -0,0 +1,201 @@
+        }
+    }
+
+    NSString *segmentAnonymousId = [self.analytics getAnonymousId];
+    if (!trackKnownUsers && self.userTraits.count > 0) {
+        [self.client track:payload.event userId:segmentAnonymousId attributes:self.userTraits eventTags:payload.properties];
+        SEGLog(@"[optimizely track:%@ userId:%@ attributes:%@ eventTags:%@]", payload.event, segmentAnonymousId, self.userTraits, payload.properties);
+    } else {
+        [self.client track:payload.event userId:segmentAnonymousId eventTags:payload.properties];
+        SEGLog(@"[optimizely track:%@ userId:%@ eventTags:%@]", payload.event, segmentAnonymousId, payload.properties);
+    }
+}
+
+- (void)reset
+{
+    if ([self.manager getOptimizely] == nil) {
+        return;
+    } else {
+        [[NSNotificationCenter defaultCenter] removeObserver:self.observer];
+        SEGLog(@"[NSNotificationCenter defaultCenter] removeObserver:%@", self.observer);
+    }
+}
+
+#pragma mark - Experiment Viewed
+
+- (void)experimentDidGetViewed:(NSNotification *)notification
+{
+    OPTLYExperiment *experiment = notification.userInfo[OptimizelyNotificationsUserDictionaryExperimentKey];
+    OPTLYVariation *variation = notification.userInfo[OptimizelyNotificationsUserDictionaryVariationKey];
+
+    NSMutableDictionary *properties = [[NSMutableDictionary alloc] init];
+    properties[@"experimentId"] = [experiment experimentId];
+    properties[@"experimentName"] = [experiment experimentKey];
+    properties[@"variationId"] = [variation variationId];
+    properties[@"variationName"] = [variation variationKey];
+
+    if ([(NSNumber *)[self.settings objectForKey:@"nonInteraction"] boolValue]) {
+        properties[@"nonInteraction"] = @1;
+    }
+
+    // Trigger event as per our spec https://segment.com/docs/spec/ab-testing/
+    [self.analytics track:@"Experiment Viewed" properties:properties options:@{
+        @"integrations" : @{
+            @"Optimizely X" : @NO
+        }
+    }];
+}
+
+#pragma mark - Private - Queueing
+
+- (void)enqueueAction:(SEGTrackPayload *)payload
+{
+    [self dispatchBackground:^{
+        {
+            SEGLog(@"%@ Optimizely not initialized. Enqueueing action: %@", self, payload);
+            @try {
+        if (self.queue.count > 100) {
+            // Remove the oldest element.
+            [self.queue removeObjectAtIndex:0];
+            SEGLog(@"%@ removeObjectAtIndex: 0", self.queue);
+        }
+        [self setupTimer];
+        [self.queue addObject:payload];
+        SEGLog(@"SEGOptimizelyX background queue length %i", self.queue.count);
+            }
+            @catch (NSException *exception) {
+        SEGLog(@"%@ Error writing payload: %@", self, exception);
+            }
+}
+}];
+}
+
+- (NSMutableArray *)queue
+{
+    if (!_queue) {
+        _queue = [NSMutableArray arrayWithCapacity:100];
+    }
+
+    return _queue;
+}
+
+- (void)setupTimer
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Does this need to be on the main queue? `isOptimizelyInitialized` is internally running on _backgroundQueue anyway.

Reply:
Yes it does have to be dispatched to the main queue